kubernetes/nodecollection.go

```diff
@@ -86,37 +87,94 @@ func (cns ClientsetNodeSource) NodeAddress(node *v1.Node) (string, int32, error)
 	return "", 0, fmt.Errorf("Could not find internal IP address for node %s ", node.Name)
 }
 
-// Endpoint an enumeration representing the various metrics endpoints
-type Endpoint uint8
+// Connection is a bitmask that describes the manner(s) in which
+// the agent can connect to an endpoint
+type Connection uint8
+
+const (
+	Unreachable Connection = 0
+	Direct      Connection = 1 << iota
+	Proxy
+)
+
+func (c Connection) HasMethod(method Connection) bool { return c&method != 0 }
+func (c *Connection) AddMethod(method Connection)     { *c |= method }
+func (c *Connection) ClearMethod(method Connection)   { *c &= ^method }
+func (c *Connection) ToggleMethod(method Connection)  { *c ^= method }
+
+func (c Connection) String() string {
+	if c == Unreachable {
+		return unreachable
+	}
+	var options []string
+	if c.HasMethod(Proxy) {
+		options = append(options, proxy)
+	}
+	if c.HasMethod(Direct) {
+		options = append(options, direct)
+	}
+	return strings.Join(options, ",")
+}
+
+type ConnectionMethod struct {
+	ConnType     Connection
+	API          nodeAPI
+	client       raw.Client
+	FriendlyName string
+}
+
+// Endpoint represents the various metrics endpoints we hit
+type Endpoint string
 
 const (
 	// NodeStatsSummaryEndpoint the /stats/summary endpoint
-	NodeStatsSummaryEndpoint Endpoint = iota
+	NodeStatsSummaryEndpoint Endpoint = "/stats/summary"
 
 	// NodeContainerEndpoint the /stats/container endpoint
-	NodeContainerEndpoint
+	NodeContainerEndpoint Endpoint = "/stats/container"
 
-	// NodeCadvisorEndpoint the /metrics/cadvisor/prometheus endpoint
-	NodeCadvisorEndpoint
+	// NodeCadvisorEndpoint the /metrics/cadvisor endpoint
+	NodeCadvisorEndpoint Endpoint = "/metrics/cadvisor"
 )
 
 // EndpointMask a map representing the currently active endpoints.
 // The keys of the map are the currently active endpoints.
-type EndpointMask map[Endpoint]struct{}
+type EndpointMask map[Endpoint]Connection
 
 // SetAvailable sets an endpoint availability state according to the supplied boolean
-func (m EndpointMask) SetAvailable(endpoint Endpoint, available bool) {
+func (m EndpointMask) SetAvailable(endpoint Endpoint, method Connection, available bool) {
+	e := m[endpoint]
 	if available {
-		m[endpoint] = struct{}{}
+		e.AddMethod(method)
 	} else {
-		delete(m, endpoint)
+		e.ClearMethod(method)
 	}
+	m[endpoint] = e
 }
 
-// Available gets the availability of an endpoint
-func (m EndpointMask) Available(endpoint Endpoint) bool {
-	_, ok := m[endpoint]
-	return ok
+// Available gets the availability of an endpoint for the specified connection method
+func (m EndpointMask) Available(endpoint Endpoint, method Connection) bool {
+	e, ok := m[endpoint]
+	if !ok {
+		return false
+	}
+	return e.HasMethod(method)
```

Comment:
I think this can be the same as the other methods i.e. don't check the `ok` flag, just make use of the Connection default value if it isn't set.